ulib/FStar.Matrix.fst

```diff
@@ -0,0 +1,839 @@
+  () in matrix_equiv_from_proof eq lhs rhs aux
+ 
+let matrix_mul_unit_row_lemma #c #eq m (add mul: CE.cm c eq) (i: under m)
+  : Lemma ((row (matrix_mul_unit add mul m) i 
+             == (SB.create i add.unit) `SB.append` 
+                ((SB.create 1 mul.unit) `SB.append` (SB.create (m-i-1) add.unit))) /\
+           (row (matrix_mul_unit add mul m) i 
+             == ((SB.create i add.unit) `SB.append` (SB.create 1 mul.unit)) `SB.append`
+                (SB.create (m-i-1) add.unit))) =  
+  SB.lemma_eq_elim ((SB.create i add.unit `SB.append` SB.create 1 mul.unit) 
+                   `SB.append` (SB.create (m-i-1) add.unit))
+                   (row (matrix_mul_unit add mul m) i); 
+  SB.lemma_eq_elim ((SB.create i add.unit) `SB.append` 
+                   (SB.create 1 mul.unit `SB.append` SB.create (m-i-1) add.unit))
+                   (row (matrix_mul_unit add mul m) i)
+                   
+let matrix_mul_unit_col_lemma #c #eq m (add mul: CE.cm c eq) (i: under m)
+  : Lemma ((col (matrix_mul_unit add mul m) i 
+             == (SB.create i add.unit) `SB.append` 
+                ((SB.create 1 mul.unit) `SB.append` (SB.create (m-i-1) add.unit))) /\
+           (col (matrix_mul_unit add mul m) i == 
+                ((SB.create i add.unit) `SB.append` (SB.create 1 mul.unit)) `SB.append`
+                (SB.create (m-i-1) add.unit))) =  
+  SB.lemma_eq_elim ((SB.create i add.unit `SB.append` SB.create 1 mul.unit) 
+                   `SB.append` (SB.create (m-i-1) add.unit))
+                   (col (matrix_mul_unit add mul m) i); 
+  SB.lemma_eq_elim ((SB.create i add.unit) `SB.append` 
+                   (SB.create 1 mul.unit `SB.append` SB.create (m-i-1) add.unit))
+                   (col (matrix_mul_unit add mul m) i)
+ 
+let seq_of_products_zeroes_lemma #c #eq #m (mul: CE.cm c eq) 
+                                 (z: c{is_absorber z mul})
+                                 (s: SB.seq c{SB.length s == m})
+  : Lemma (ensures (eq_of_seq eq (seq_of_products mul (SB.create m z) s) (SB.create m z))) 
+  = eq_of_seq_from_element_equality eq (seq_of_products mul (SB.create m z) s) (SB.create m z) 
+
+let rec foldm_snoc_zero_lemma #c #eq (add: CE.cm c eq) (zeroes: SB.seq c)
+  : Lemma (requires (forall (i: under (SB.length zeroes)). SB.index zeroes i `eq.eq` add.unit))
+          (ensures eq.eq (SP.foldm_snoc add zeroes) add.unit) 
+          (decreases SB.length zeroes) =
+  if (SB.length zeroes < 1) then begin
+    assert_norm (SP.foldm_snoc add zeroes == add.unit);
+    eq.reflexivity add.unit    
+  end else
+  let liat, last = SProp.un_snoc zeroes in
+  foldm_snoc_zero_lemma add liat;
+  add.congruence last (SP.foldm_snoc add liat) add.unit add.unit;
+  add.identity add.unit;
+  foldm_snoc_last add zeroes;
+  eq.transitivity (SP.foldm_snoc add zeroes)                                
+                  (add.mult add.unit add.unit)
+                  add.unit
+ 
+#push-options "--z3rlimit 150 --fuel 1 --ifuel 0" 
```

Comment:
Same here ... this is also too brittle to maintain

Reply:
Similarly, on my working machine this one passes at 
`#push-options "--z3rlimit 20 --fuel 1 --ifuel 0" `

I'll try to improve both, not sure if I'll succeed tho :)

Reply:
I believe I did it :)